user-manual/src/config/kdl.md

```diff
@@ -255,6 +264,121 @@ Filters at this stage are the earliest. Currently supported filters:
     * The given header will be added or replaced to `VALUE`
 
+### `services.$NAME.rate-limiting`
+
+This section contains the configuration for rate limiting rules.
+
+Rate limiting rules are used to limit the total number of requests made by downstream clients,
+based on various criteria.
+
+Note that Rate limiting is on a **per service** basis, services do not share rate limiting
+information.
+
+This section is optional.
+
+Example:
+
+```
+rate-limiting {
+    timeout millis=100
+
+    rule kind="source-ip" \
+        max-buckets=4000 tokens-per-bucket=10 refill-qty=1 refill-rate-ms=10
+
+    rule kind="uri" pattern="static/.*" \
+        max-buckets=2000 tokens-per-bucket=20 refill-qty=5 refill-rate-ms=1
+}
+```
+
+#### `services.$NAME.rate-limiting.timeout`
+
+The `timeout` parameter is used to set the total timeout for acquiring all rate limiting tokens.
+
+If acquiring applicable rate limiting tokens takes longer than this time, the request will not be
+forwarded and will respond with a 429 error.
+
+This parameter is mandatory if the `rate-limiting` section is present.
+
+This is specified in the form:
+
+`timeout millis=TIME`, where `TIME` is an unsigned integer
+
+**Implementation Note**: The rate limiting timeout is a temporary implementation detail to limit
+requests from waiting "too long" to obtain their tokens. In the future, it is planned to modify
+the leaky bucket implementation to instead set an upper limit on the maximum "token debt", or
+how many requests are waiting for a token. Instead of waiting and timing out, requests will instead
+be given immediate feedback that the rate limiting is overcongested, and return a 429 error immediately,
+instead of after a given timeout.
+
+When this change occurs, the `timeout` parameter will be deprecated, and replaced with a `max-token-debt`
+parameter instead.
+
+#### `services.$NAME.rate-limiting.rule`
+
+Rules are used to specify rate limiting parameters, and applicability of rules to a given request.
+
+##### Leaky Buckets
+
+Rate limiting in River uses a [Leaky Bucket] model for determining whether a request can be served
+immediately, or if it should be delayed (or rejected). For a given rule, a "bucket" of "tokens"
+is created, where one "token" is required for each request.
+
+The bucket for a rule starts with a configurable `tokens-per-bucket` number. When a request arrives,
+it attempts to take one token from the bucket. If one is available, it is served immediately. Otherwise,
+the request waits in a first-in, first-out order for a token to become available.
+
+The bucket is refilled at a configurable rate, specified by `refill-rate-ms`, and adds a configurable
+number of tokens specified by `refill-qty`. The number of tokens in the bucket will never exceed the
+initial `tokens-per-bucket` number.
+
+Once a refill occurs, requests may become ready if a token becomes available.
```

Comment:
```suggestion
Once the bucket is refilled, queued requests may take a token and begin processing.
```

(wordsmithing only, please check for accuracy)

Reply:
I think I'll decline this specific wording, because I want to make it clear that we don't have to wait until the bucket is FULLY refilled, e.g. if it has room for 10 tokens, and refills 2 tokens every millisecond, we'll serve two new requests every millisecond, we don't have to wait 5 milliseconds for the bucket to fully make it back to 10.

I'll take a look at rewording this line though!

Reply:
This went away when removing the concept of delay and timeouts